query-node/joystream-query-node/schema.graphql

```diff
@@ -1,21 +1,212 @@
+	about: String
 
-  # Blocknumber when member was registered
-  registeredAtBlock: BigInt!
+	"Blocknumber when member was registered"
+	registeredAtBlock: BigInt!
 
-  # Member's controller account id.
-  controllerAccount: Bytes!
+	"Member's controller account id"
+	controllerAccount: Bytes!
 
-  # Member's root account id
-  rootAccount: Bytes!
+	"Member's root account id"
+	rootAccount: Bytes!
+}
+
+"""
+It represents a conceptual category, similar - but not identical,
+to tables in relational databases and classes in object-oriented programming
+Examples of classes in our example could be:
+- Person: represents the concept of a person.
+- Place: represents the concept of a place.
+- PersonLivedInPlace: represents the concept that a person lived in a place.
+"""
+type Class @entity {
+	"Runtime identifier for class"
+	classId: BigInt!
+
+	"Name of class"
+	name: String!
+
+	"Description of class"
+	description: String!
+
+	"The maximum number of entities which can be created"
+	maximumEntitiesCount: BigInt!
+
+	"The current number of entities which exist"
+	currentNumberOfEntities: BigInt!
+
+	"How many entities a given controller may create at most"
+	defaultEntityCreationVoucherUpperBound: BigInt!
+
+	"For this permission, the individual member is allowed to create the entity and become controller"
+	anyMemberCanCreate: Boolean
+
+	"Whether to prevent everyone from creating an entity"
+	entityCreationBlocked: Boolean!
+
+	"Whether to prevent everyone from updating entity properties"
+	allEntityPropertyValuesLocked: Boolean!
+
+	"All properties that have been used on this class across different class schemas"
+	properties: [Property!]
+
+	"All schemas that are available for this class, think v0.0 Person, v.1.0 Person, etc."
+	schemas: [Schema!] @derivedFrom(field: "class")
+
+	"Current class maintainer curator groups"
+	maintainers: [CuratorGroup!]
+
+	"All vouchers for this class"
+	vouchers: [EntityCreationVoucher!] @derivedFrom(field: "class")
+}
+
+enum PropertyType {
+	BOOL
+	UINT16
+	UINT32
+	UINT64
+	INT16
+	INT32
+	INT64
+	TEXT
+	REFERENCE
+	VECBOOL
+	VECUINT16
+	VECUINT32
+	VECUINT64
+	VECINT16
+	VECINT32
+	VECINT64
+	VECTEXT
+	VECREFERENCE
+}
+
+"`Property` representation, related to a given `Class`"
+type Property @entity {
+	"Runtime identifier for property in class"
+	propertyID: BigInt!
+
+	"Type of `Property`"
+	propertyType: PropertyType!
+
+	"If property value can be skipped, when adding entity schema support"
+	required: Boolean!
+
+	"Used to enforce uniqeness of a property across all entities that have this property"
+	unique: Boolean!
+
+	"Name of property"
+	name: String!
+
+	"Description of property"
+	description: String!
+
+	"Property is locked from maintainer"
+	isLockedFromMaintainer: Boolean!
+
+	"Property is locked from controller"
+	isLockedFromController: Boolean!
+
+	"Upper bound for vector length when property is a vector"
+	maximumLength: BigInt
```

Comment:
Shouldn't we create a separate schema for `VecPropertyType` ?https://github.com/iorveth/joystream/blob/content_directory_second_try/runtime-modules/content-directory/src/schema.rs#L91

Afterwards we possibly can utilize union type to choose between `Type` or `VecType`, can't we? 
It`s seems much clearer, IMHO. https://graphql.org/learn/schema/#union-types

```
"Optional field, it must reference a class then PropertyType is Reference"
	referencedClass: Class
	"Optional field, Only will be set when PropertyType is Reference"
	sameControllerStatus: Boolean
```

This fields should go to `VecPropertyType`

Reply:
Yeah, agree with that. We have `maximumLength` field when the property type is `Vector` and [`referencedClass`, `sameControllerStatus`] fields when the property type is `Reference`. So that means we have two options if we use union type (We have `variant` directive that can be used) ie:

```
type ReferenceTypeData @variant {
  referencedClass: Class
  sameControllerStatus: Boolean
}

type VectorTypeData @variant {
  maximumLength: BigInt
}
union PropertyData =  ReferenceTypeData | VectorTypeData
type Property @entity {
   #...
   propertyData: PropertyData
```

But a property type has 3 variants actually `Reference`, `Vector` and (`Int*, `Text`). If you notice there is no extra field exists when property type is (`Int*, `Text`). I mean if i could have at least a field for (`Int*, `Text`) ie:

```
type TextIntTypeData @variant {
  someField: SomeType
}
```
then it would be okey to do that. I hope I could explain it :)

Reply:
Oh, i see, type system is much less flexible here. 
Thank you for the clarification.